File: sample-chat/sample-chat/ReplaceSegue/ReplaceSegue.m
//
//  ReplaceSegue.m
//  sample-chat
//
//  Created by Anton Sokolchenko on 3/23/16.
//  Copyright © 2016 Quickblox. All rights reserved.
//

#import "ReplaceSegue.h"

@implementation ReplaceSegue

-(void)perform {
	
	// Grab Variables for readability
	UIViewController *sourceViewController = (UIViewController*)[self sourceViewController];
	UIViewController *destinationController = (UIViewController*)[self destinationViewController];
	UINavigationController *navigationController = sourceViewController.navigationController;
	
	if (navigationController.viewControllers != nil) {
		
		// Get a changeable copy of the stack
		NSMutableArray *controllerStack = navigationController.viewControllers.mutableCopy;
		
		// Replace the source controller with the destination controller, wherever the source may be
		[controllerStack replaceObjectAtIndex:[controllerStack indexOfObject:sourceViewController] withObject:destinationController];
		
		// Assign the updated stack with animation
		[navigationController setViewControllers:controllerStack animated:YES];
	}
}

@end

File: sample-chat-swift/sample-chat-swift/Classes/ReplaceSegue/ReplaceSegue.swift
//
//  ReplaceSegue.swift
//  sample-chat-swift
//
//  Created by Anton Sokolchenko on 3/18/16.
//  Copyright © 2016 quickblox. All rights reserved.
//

import Foundation

class ReplaceSegue : UIStoryboardSegue {
	
	override func perform() {
		let sourceViewController = self.source
		let destinationViewController = self.destination
		
		let navigationController = sourceViewController.navigationController
		
		navigationController?.pushViewController(destinationViewController, animated: false)
        
		guard var mutableVC = navigationController?.viewControllers else {
			print("Error: no view controllers")
			return
		}
		
		guard let sourceViewControllerIndex = mutableVC.index(of: sourceViewController) else {
			print("Error: no index for source view controller")
			return
		}

		mutableVC.remove(at: sourceViewControllerIndex)
		
		navigationController?.setViewControllers(mutableVC, animated: true)
	}
	
}
